File: coin-change.py
class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        if amount == 0:
            return 0

        if min(coins) > amount:
            return -1

        amount_list = []
        amount_list = [-1] * (amount + 1)
        amount_list[0] = 0

        for i in range(1, amount + 1):
            for c in coins:
                if c <= i:
                    prev = amount_list[i - c]

                    if (prev != -1 and
                       (amount_list[i] == -1 or prev + 1 < amount_list[i])):
                        amount_list[i] = prev + 1

        return amount_list[amount]



if __name__ == '__main__':
    s = Solution()

    coins_list = [
            [1, 2, 5],
            [2],
            [186,419,83,408],
    ]

    amount_list = [
            11,
            3,
            6249,
    ]

    for i in range(len(coins_list)):
        print(s.coinChange(coins_list[i], amount_list[i]))

File: coin-change.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
var coinChange = function(coins, amount) {
    var amountList = [];
    var i, j;

    amountList[0] = 0;

    for (i = 0; i < amount + 1; i++) {
        for (j = 0; j < coins.length; j++) {
            if (coins[j] <= i) {
                var prev = amountList[i - coins[j]];

                if (prev !== undefined && (!amountList[i] || prev + 1 < amountList[i])) {
                    amountList[i] = prev + 1;
                }
            }
        }
    }

    return amountList[amount] !== undefined ? amountList[amount] : -1;
};

var main = function() {
  coins = [1, 2, 5];
  amount = 11;

  console.log(coinChange(coins, amount));
}

if (require.main === module) {
  main();
}
